programs/mango-v4/src/health/cache.rs

```diff
@@ -377,35 +471,135 @@ pub struct HealthCache {
 
 impl HealthCache {
     pub fn health(&self, health_type: HealthType) -> I80F48 {
+        let token_balances = self.effective_token_balances(health_type);
         let mut health = I80F48::ZERO;
         let sum = |contrib| {
             health += contrib;
         };
-        self.health_sum(health_type, sum);
+        self.health_sum(health_type, sum, &token_balances);
         health
     }
 
-    /// Sum of only the positive health components (assets) and
-    /// sum of absolute values of all negative health components (liabs, always >= 0)
-    pub fn health_assets_and_liabs(&self, health_type: HealthType) -> (I80F48, I80F48) {
-        let mut assets = I80F48::ZERO;
-        let mut liabs = I80F48::ZERO;
-        let sum = |contrib| {
-            if contrib > 0 {
-                assets += contrib;
+    pub fn health_assets_and_liabs_stable_assets(
+        &self,
+        health_type: HealthType,
+    ) -> (I80F48, I80F48) {
+        self.health_assets_and_liabs(health_type, true)
+    }
+    pub fn health_assets_and_liabs_stable_liabs(
```

Comment:
hmm this seems to be only used in the group level deposit limit feature which is disable for now, I was wondering why these family of functions are in program, anyway I guess we can keep them here